### PR TITLE
network: introduce L2TP tunnel

### PR DESCRIPTION
**Pull request by yuwata**
This includes #11908. Changes from #11908 are
- Add `NETDEV_CREATE_AFTER_CONFIGURED` creation type.
- use `NetworkConfigSection` to manage `[L2TPSession]` section.
- move `Layer2SpecificHeader=` from `[L2TP]` to `[L2TPSession]`.
- make `Local=` support an empty string, and special values `auto`, `static`, and `dynamic`.
- all comments by @poettering in #11908 are addressed.
- many tiny cleanups.

Known issue(?)
- After creating l2tp tunnel or session, sd-netlink receives unsupported reply.
```
Mar 12 11:32:56 systemd-networkd[16579]: enp0s20u1: Enslaving by 'l2tp-test'
Mar 12 11:32:56 systemd-networkd[16579]: l2tp-test: Local address 192.168.42.153 acquired.
Mar 12 11:32:56 systemd-networkd[16579]: l2tp-test: L2TP tunnel created
Mar 12 11:32:56 systemd-networkd[16579]: sd-netlink: ignored message with unknown type: 2
Mar 12 11:32:56 systemd-networkd[16579]: sd-netlink: ignored message with unknown type: 2
Mar 12 11:32:56 systemd-networkd[16579]: l2tp-test: L2TP session test1 created
Mar 12 11:32:56 systemd-networkd[16579]: sd-netlink: ignored message with unknown type: 2
Mar 12 11:32:56 systemd-networkd[16579]: Got message type=method_return sender=:1.351 destination=:1.350 path=n/a interface=n/a member=n/a cookie=6 reply_cookie=14 signature=n/a error-name=n/a error-message=n/a
Mar 12 11:32:56 systemd-networkd[16579]: test1: New device has no master, continuing without
Mar 12 11:32:56 systemd-networkd[16579]: test1: Flags change: +MULTICAST +BROADCAST
Mar 12 11:32:56 systemd-networkd[16579]: test1: Link 7 added
Mar 12 11:32:56 systemd-networkd[16579]: test1: link pending udev initialization...
Mar 12 11:32:56 systemd-networkd[16579]: test1: Saved original MTU: 1458
Mar 12 11:32:56 systemd-networkd[16579]: sd-netlink: ignored message with unknown type: 2
```

This PR depends on #11861, so please do not merge. No test is included yet.

cc @ssahani.

### CODE REVIEW

**Comment by ssahani** [comment]
thanks for putting it here @yuwata .

```After creating l2tp tunnel or session, sd-netlink receives unsupported reply.``` Yes I seen these . Guess safely can be ignored.

**Comment by yuwata** [comment]
A simple test is added. But it may be flaky...

**Comment by yuwata** [comment]
> Guess safely can be ignored.

@ssahani Thanks. I hope so. But not sure :-p

**Comment by yuwata** [comment]
Hmm, the log seems truncated at least when I see it from my cellphone, but at least `test_l2tp` fails. I will take a look.

**Comment by yuwata** [comment]
Hmm, the test also randomly fails at local. I guess, but not sure, may be we need to wait for the tunnel to be created before creating sessions.

**Comment by ssahani** [comment]
> Hmm, the test also randomly fails at local. I guess, but not sure, may be we need to wait for the tunnel to be created before creating sessions.

How about putting a sleep ?
```
time.sleep(1)

self.assertTrue(self.link_exits('test1'))
time.sleep(3)
self.assertTrue(self.link_exits('l2tp-ses1'))
self.assertTrue(self.link_exits('l2tp-ses2'))

```￼

**Comment by yuwata** [comment]
Maybe no. If it fails, the log of network does not show any response about the sessions to be created.

**Comment by ssahani** [comment]
oopps sorry . I just overlooked. you need to remove the code. It does not any create any link . This only can be only ip l2tp show

```
self.assertTrue(self.link_exits('test1'))
self.assertTrue(self.link_exits('l2tp-ses1'))
self.assertTrue(self.link_exits('l2tp-ses2')
```


ip l2tp show tunnel
```
Tunnel 3000, encap UDP
  From 192.168.137.1 to 192.168.137.2
  Peer tunnel 4000
  UDP source / dest ports: 5000/6000
  UDP checksum: disabled
```

File: src/network/netdev/l2tp-tunnel.c [hunk 1]
```diff
@@ -0,0 +1,586 @@
...
+
+        if (!session->name)
+                return log_netdev_error_errno(netdev, SYNTHETIC_ERRNO(EINVAL),
+                                              "%s: L2TP session without name configured. "
+                                              "Ignoring [L2TPSession] section from line %u",
+                                              session->section->filename, session->section->line);
+
+        if (session->session_id == 0 || session->peer_session_id == 0)
+                return log_netdev_error_errno(netdev, SYNTHETIC_ERRNO(EINVAL),
+                                              "%s: L2TP session without name configured. "
```
Comment (poettering) [inline src/network/netdev/l2tp-tunnel.c]: without *session IDs*

File: src/systemd/sd-netlink.h [hunk 2]
```diff
@@ -34,7 +34,13 @@ typedef struct sd_netlink sd_netlink;
 typedef struct sd_genl_socket sd_genl_socket;
 typedef struct sd_netlink_message sd_netlink_message;
 typedef struct sd_netlink_slot sd_netlink_slot;
-typedef enum {SD_GENL_ID_CTRL, SD_GENL_WIREGUARD, SD_GENL_FOU} sd_genl_family;
+
+typedef enum sd_gen_family {
+              SD_GENL_ID_CTRL,
```
Comment (poettering) [inline src/systemd/sd-netlink.h]: that's 16ch, no? should be 8ch indenting

File: src/network/netdev/netdev.c [hunk 3]
```diff
@@ -35,6 +35,7 @@
 #include "netdev/wireguard.h"
 #include "netdev/netdevsim.h"
 #include "netdev/fou-tunnel.h"
+#include "netdev/l2tp-tunnel.h"
```
Comment (poettering) [inline src/network/netdev/netdev.c]: these should probably be sorted again

**Comment by yuwata** [comment]
The comments by @poettering are addressed, but test is not fixed, so CentOS CI may randomly fail.

**Comment by yuwata** [comment]
Updated.
- handle previously unknown replies.
- wait for reply from kernel before creating sessions

At least on my laptop, now the test always succeeds.

**Comment by poettering** [comment]
please rebase, now that #11861 has been merged!

**Comment by yuwata** [comment]
Hmm, still the new test fails on CentOS. The link existence tests are fine on my laptop, but may not be supported by old kernels. I will check that later.

**Comment by yuwata** [comment]
Rebased, and commented out the link existence checks, as @ssahani commented at https://github.com/systemd/systemd/pull/11968#issuecomment-472006988. Though, still I have no idea why the tests pass on my laptop but not on CentOS CI. Anyway, let's wait for the CI results.

**Comment by mrc0mmand** [comment]
@yuwata I think you keep hitting the 15 minutes timeout in [CentOS CI](https://github.com/systemd/systemd-centos-ci/blob/master/agent/testsuite.sh#L92). I guess bumping it a little should help.

**Comment by mrc0mmand** [comment]
*Ninja-commited* https://github.com/mrc0mmand/systemd-centos-ci/commit/a024c06253c60a1adf57525f620c409e7860d486 and restarted the CentOS CI job. Let's see if it was the root cause.

**Comment by yuwata** [comment]
@mrc0mmand Thanks. But L2TP related tests run and fail...
```
test_l2tp_ip (__main__.NetworkdL2TPTests) ... 

FAIL 
test_l2tp_udp (__main__.NetworkdL2TPTests) ... 
Tunnel 10, encap UDP
 From 192.168.30.100 to 1.2.3.4
 Peer tunnel 11
 UDP source / dest ports: 3000/4000
 UDP checksum: enabled

FAIL
```

**Comment by mrc0mmand** [comment]
> 
> 
> @mrc0mmand Thanks. But L2TP related tests run and fail...
> 
> ```
> test_l2tp_ip (__main__.NetworkdL2TPTests) ... 
> 
> FAIL 
> test_l2tp_udp (__main__.NetworkdL2TPTests) ... 
> Tunnel 10, encap UDP
>  From 192.168.30.100 to 1.2.3.4
>  Peer tunnel 11
>  UDP source / dest ports: 3000/4000
>  UDP checksum: enabled
> 
> FAIL
> ```

Ah, I completely overlooked this fail. However, the test was triggering the timeout as well, so hopefully it resolves at least one of the issues.

**Comment by ssahani** [comment]
Both test cases passed in my laptop
```
test_l2tp_ip (__main__.NetworkdL2TPTests) ... 


Tunnel 10, encap IP
  From 192.168.30.100 to 192.168.30.101
  Peer tunnel 12
Session 27 in tunnel 10
  Peer session 28, tunnel 12
  interface name: l2tp-ses4
  offset 0, peer offset 0

Session 25 in tunnel 10
  Peer session 26, tunnel 12
  interface name: l2tp-ses3
  offset 0, peer offset 0
Session 27 in tunnel 10
  Peer session 28, tunnel 12
  interface name: l2tp-ses4
  offset 0, peer offset 0

Session 25 in tunnel 10
  Peer session 26, tunnel 12
  interface name: l2tp-ses3
  offset 0, peer offset 0
ok
test_l2tp_udp (__main__.NetworkdL2TPTests) ... 
Tunnel 10, encap UDP
  From 192.168.30.100 to 1.2.3.4
  Peer tunnel 11
  UDP source / dest ports: 3000/4000
  UDP checksum: enabled
Session 17 in tunnel 10
  Peer session 18, tunnel 11
  interface name: l2tp-ses2
  offset 0, peer offset 0

Session 15 in tunnel 10
  Peer session 16, tunnel 11
  interface name: l2tp-ses1
  offset 0, peer offset 0
Session 17 in tunnel 10
  Peer session 18, tunnel 11
  interface name: l2tp-ses2
  offset 0, peer offset 0

Session 15 in tunnel 10
  Peer session 16, tunnel 11
  interface name: l2tp-ses1
  offset 0, peer offset 0
ok

----------------------------------------------------------------------
Ran 2 tests in 18.576s

OK

```

@yuwata Please add  ```def expectedFailureIfERSPANModuleIsNotAvailable():``` similar here to skip it .
Just see 

```
     site-A:# ip l2tp add tunnel tunnel_id 3000 peer_tunnel_id 4000 \
                  encap udp local 1.2.3.4 remote 5.6.7.8 \
                  udp_sport 5000 udp_dport 6000 
```
success if yes then continue else skip the test case 

or you can do a modprobe for the below modules

```
[root@Arch test-network]# lsmod | grep l2tp
l2tp_eth               20480  0
l2tp_ip                24576  0
l2tp_netlink           28672  1 l2tp_eth
l2tp_core              36864  3 l2tp_eth,l2tp_ip,l2tp_netlink
ip6_udp_tunnel         16384  1 l2tp_core
udp_tunnel             16384  1 l2tp_core


```

**Comment by mrc0mmand** [comment]
@yuwata @ssahani would it make sense to run the `systemd-networkd-tests.py` as part of the CentOS CI Vagrant job (i.e. in a VM with Arch), so we could test the newer features as well, or they require a baremetal machine?

**Comment by ssahani** [comment]
That would be nice. Thanks

**Comment by yuwata** [comment]
@ssahani and @mrc0mmand Thank you for your help. I added module existence tests.

For the UDP case, tunnel is successfully created but no session exists. So, I guess `l2tp_eth` module may not exist.
For the IP case, tunnel cannot be created. So, maybe `l2tp_ip` module does not exist.

**Comment by yuwata** [comment]
Yey! CentOS CI is now green!

Interestingly, the newly added tests are ok, not in expected failure. I expect that relevant modules are automatically loaded when a `l2tp` related netlink message is sent to kernel. But maybe the kernel used in the CentOS has a bug that some modules are not loaded, or the message is processed before loading modules finished. And I guess the bug is already fixed in the recent kernel at least used in Fedora29 (not sure, but by https://github.com/torvalds/linux/commit/6b9f34239b00e6956a267abed2bc559ede556ad6?).

Anyway, now the modules are always loaded before starting the tests. So, it works. Yey!

@poettering and @keszybz PTAL.

**Comment by ssahani** [comment]
> Interestingly, the newly added tests are ok, not in expected failure. I expect that relevant modules are automatically loaded when a `l2tp` related netlink message is sent to kernel. But maybe the kernel used in the CentOS has a bug that some modules are not loaded, or the message is processed before loading modules finished. And I guess the bug is already fixed in the recent kernel at least used in Fedora29 (not sure, but by [torvalds/linux@6b9f342](https://github.com/torvalds/linux/commit/6b9f34239b00e6956a267abed2bc559ede556ad6)?).
> 
> Anyway, now the modules are always loaded before starting the tests. So, it works. Yey!
This line is responsible for  autoloading.

https://elixir.bootlin.com/linux/v4.2/source/net/l2tp/l2tp_netlink.c#L999
https://lists.openwall.net/netdev/2012/05/29/53

```
MODULE_ALIAS_GENL_FAMILY("l2tp")
```

**Comment by yuwata** [comment]
The patch is included in v3.5 https://github.com/torvalds/linux/commit/e9412c37082b5c932e83364aaed0c38c2ce33acb...

**Comment by poettering** [comment]
OK, I'd lie if I'd claim that I fully understood l2tp and the semntics of its API, but the code looks superficially great and the tests pass, hence let's merge!

**Pull request merged**